src/EditorFeatures/Core/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs

```diff
@@ -0,0 +1,171 @@
+        {
+            var definition = await targetSymbol.ToClassifiedDefinitionItemAsync(
+                solution,
+                isPrimary: true,
+                includeHiddenLocations: false,
+                FindReferencesSearchOptions.Default,
+                cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            return new InheritanceTargetItem(
+                inheritanceRelationshipWithOriginalMember,
+                definition);
+        }
+
+        private static async Task<InheritanceMemberItem> CreateInheritanceMemberInfoForMemberAsync(
+            Solution solution,
+            ISymbol memberSymbol,
+            int lineNumber,
+            ImmutableArray<ISymbol> implementingMembers,
+            ImmutableArray<ISymbol> implementedMembers,
+            ImmutableArray<ISymbol> overridenMembers,
+            ImmutableArray<ISymbol> overridingMembers,
+            CancellationToken cancellationToken)
+        {
+            var implementingMemberItems = await implementingMembers
+                .SelectAsArrayAsync(symbol => CreateInheritanceItemAsync(solution, symbol, InheritanceRelationship.Implementing, cancellationToken)).ConfigureAwait(false);
+            var implementedMemberItems = await implementedMembers
+                .SelectAsArrayAsync(symbol => CreateInheritanceItemAsync(solution, symbol, InheritanceRelationship.Implemented, cancellationToken)).ConfigureAwait(false);
+            var overridenMemberItems = await overridenMembers
+                .SelectAsArrayAsync(symbol => CreateInheritanceItemAsync(solution, symbol, InheritanceRelationship.Overriden, cancellationToken)).ConfigureAwait(false);
+            var overridingMemberItems = await overridingMembers
+                .SelectAsArrayAsync(symbol => CreateInheritanceItemAsync(solution, symbol, InheritanceRelationship.Overriding, cancellationToken)).ConfigureAwait(false);
+
+            return new InheritanceMemberItem(
+                lineNumber,
+                memberSymbol.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat),
+                memberSymbol.GetGlyph(),
+                implementingMemberItems.Concat(implementedMemberItems)
+                    .Concat(overridenMemberItems)
+                    .Concat(overridingMemberItems));
+        }
+
+        /// <summary>
+        /// For the <param name="memberSymbol"/>, get all the symbols implementing it.
+        /// </summary>
+        private static async Task<ImmutableArray<ISymbol>> GetImplementedSymbolsAsync(
+            Solution solution,
+            ISymbol memberSymbol,
+            CancellationToken cancellationToken)
+        {
+            if (memberSymbol is INamedTypeSymbol { IsSealed: false } namedTypeSymbol)
+            {
+                var derivedTypes = await GetDerivedTypesAndImplementationsAsync(solution, namedTypeSymbol, cancellationToken).ConfigureAwait(false);
+                return derivedTypes.OfType<ISymbol>().ToImmutableArray();
```

Comment:
avoid the redundant OfTYpe to move up/down the inheritance hierarchy.

Reply:
This pattern occurs in several places. I gave two proposed solutions in https://github.com/dotnet/roslyn/pull/51929#discussion_r596095977.

Reply:
Will make sure all the OfTYpe Are removed